### PR TITLE
algorithm: SegmentTree

### PR DESCRIPTION
**Pull request by changicho**
[Open in Gitpod](https://gitpod.io/from-referrer/)&nbsp;[know more](https://www.gitpod.io/docs/pull-requests/)

### Describe your change:

- [x] Add an algorithm?
- [ ] Fix a bug or typo in an existing algorithm?
- [ ] Documentation change?

### Checklist:

- [x] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Javascript/blob/master/CONTRIBUTING.md).
- [x] This pull request is all my own work -- I have not plagiarized.
- [x] I know that pull requests will not be merged if they fail the automated tests.
- [x] This PR only changes one algorithm file. To ease review, please open separate PRs for separate algorithms.
- [x] All new JavaScript files are placed inside an existing directory.
- [x] All filenames should use the UpperCamelCase (PascalCase) style. There should be no spaces in filenames.
      **Example:**`UserProfile.js` is allowed but `userprofile.js`,`Userprofile.js`,`user-Profile.js`,`userProfile.js` are not
- [x] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
- [ ] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

**Review by appgurueu — COMMENTED** [review]
Elegant & clever implementation. Can you try to add more comments?

**Comment by changicho** [comment]
> Elegant & clever implementation. Can you try to add more comments?

Sure! I add comments in new commit. Please check. Thanks.

**Comment by appgurueu** [comment]
I was thinking of comments to the implementation (e.g. how you store the binary tree in a list (same trick as heaps often use), how to calc. the parent pos. from a child pos. and vice versa).

**Comment by changicho** [comment]
> I was thinking of comments on the implementation details (e.g. how you store the binary tree in a list (same trick as heaps often use), how to calc. the parent pos. from a child pos. and vice versa).

I add comments on the implementation details in new commit!

File: Data-Structures/Tree/SegmentTree.js [hunk 1]
```diff
@@ -14,53 +14,80 @@ class SegmentTree {
...
+      tree[size + i] = arr[i]
     }
 
     // build the tree by calculating parents
-    for (let i = n - 1; i > 0; --i) {
-      this.tree[i] = this.tree[i << 1] + this.tree[(i << 1) | 1]
+    // tree's root node will contain all leaf node's sum
+    for (let i = size - 1; i > 0; --i) {
+      // current node's value is the sum of left child, right child
+      // tree[i] = tree[i * 2] + tree[i * 2 + 1]
```
Comment (appgurueu) [inline Data-Structures/Tree/SegmentTree.js:44]: this comment is quite literally redundant :sweat_smile:

**Review by appgurueu — DISMISSED** [review]

File: Data-Structures/Tree/SegmentTree.js (hunk 1 above)
Comment (changicho) [reply Data-Structures/Tree/SegmentTree.js]: 😅 OMG! I forgot to erase that line. thanks for your review

**Comment by changicho** [comment]
@raklaptudirm Could you review this PR? 🙏

**Review by appgurueu — APPROVED** [review]

**Review by raklaptudirm — APPROVED** [review]

**Pull request merged**